library/alloc/src/collections/vec_deque/iter.rs

```diff
@@ -13,38 +15,44 @@ use super::{count, wrap_index, RingSlices};
 /// [`iter`]: super::VecDeque::iter
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Iter<'a, T: 'a> {
-    ring: &'a [MaybeUninit<T>],
+    ring: *const [T],
     tail: usize,
     head: usize,
+    _marker: PhantomData<&'a T>,
 }
 
 impl<'a, T> Iter<'a, T> {
     pub(super) fn new(ring: &'a [MaybeUninit<T>], tail: usize, head: usize) -> Self {
-        Iter { ring, tail, head }
+        Iter {
+            ring: ptr::slice_from_raw_parts(ring.as_ptr().cast(), ring.len()),
```

Comment:
This could also be written as `ring as *const [MaybeUninit<T>] as *const [T]`. Not sure if the `as` cast is better than a `ptr::slice_from_raw_parts` call, but in both cases I feel like there should be a better way to write this.